pkg/storage/allocator_scorer.go

```diff
@@ -697,63 +705,74 @@ func constraintCheck(store roachpb.StoreDescriptor, constraints config.Constrain
+	if numSamples == 0 {
+		return 1.0
 	}
-	return minScore
+	return sumScore / float64(numSamples)
 }
 
-// diversityRemovalScore is the same as diversityScore, but for a node that's
-// already present in existingNodeLocalities. It works by calculating the
-// diversityScore for nodeID as if nodeID didn't already have a replica.
-// As with diversityScore, a higher score indicates that the node is a better
-// fit for the range (i.e. keeping it around is good for diversity).
+// diversityRemovalScore calculates the average diversityScore if we try to remove the replica from this node.
+// A higher score indicates that the node is a better fit for the range (i.e. keeping it around is good for diversity).
 func diversityRemovalScore(
 	nodeID roachpb.NodeID, existingNodeLocalities map[roachpb.NodeID]roachpb.Locality,
 ) float64 {
-	minScore := roachpb.MaxDiversityScore
+	var sumScore float64
+	var numSamples int
 	locality := existingNodeLocalities[nodeID]
+	// We didn't calculate the overall diversityScore for the range, because the original overall diversityScor
+	// of this range is always the same.
 	for otherNodeID, otherLocality := range existingNodeLocalities {
 		if otherNodeID == nodeID {
 			continue
 		}
-		if newScore := otherLocality.DiversityScore(locality); newScore < minScore {
-			minScore = newScore
-		}
+		newScore := locality.DiversityScore(otherLocality)
+		sumScore += newScore
+		numSamples++
 	}
-	return minScore
+	if numSamples == 0 {
+		return 0
+	}
+	return sumScore / float64(numSamples)
 }
 
-// rebalanceToDiversityScore is like diversityScore, but it returns what
-// the diversity score would be if the given store was added and one of the
-// existing stores was removed. This is equivalent to the second lowest score.
-//
-// This is useful for considering rebalancing a range that already has enough
-// replicas - it's perfectly fine to add a replica in the same locality as an
-// existing replica in such cases.
-func rebalanceToDiversityScore(
-	store roachpb.StoreDescriptor, existingNodeLocalities map[roachpb.NodeID]roachpb.Locality,
+// diversityRebalanceScore is a little complicated, we should coalesce the upcoming removing
+// replica together. A much higher score means the rebalance target store is a better fit for the range.
+func diversityRebalanceScore(
+	removeStore roachpb.StoreDescriptor,
+	rebalanceStore roachpb.StoreDescriptor,
+	existingNodeLocalities map[roachpb.NodeID]roachpb.Locality,
 ) float64 {
-	minScore := roachpb.MaxDiversityScore
-	nextMinScore := roachpb.MaxDiversityScore
-	for _, locality := range existingNodeLocalities {
-		newScore := store.Node.Locality.DiversityScore(locality)
-		if newScore < minScore {
-			nextMinScore = minScore
-			minScore = newScore
-		} else if newScore < nextMinScore {
-			nextMinScore = newScore
+	var sumScore float64
+	var numSamples int
+	for nodeID, locality := range existingNodeLocalities {
```

Comment:
I'm not convinced that it's sufficient to only count the average diversity of the new replica against all the existing ones, because the choice of `removeStore` affects the average diversity as well. Can you explain/justify why this is correct?

I kind of expected us to just do a pairwise computation over all the replicas that would be in the resulting range. That'd be much easier to understand without being slower enough to matter. I think we should switch to that unless there's some reasoning I'm not seeing. Honestly, I think we'd be best off just doing that for all three of these functions for the sake of clarity.

Reply:
For `diversityRebalanceScore`, If we have known the `removeStore`, what we need to do is to find the best rebalance target, it's just like `diversityAllocateScore`, but when we calculate the diversityScore, we just kick out the `removeStore`.